Add Lds transpose load ds read tr16

stefankoncarevic opened the pull request:
## Motivation

close: https://github.com/ROCm/rocMLIR-internal/issues/1858

## Technical Details

This change adds full support for LDS transpose load integration within both single-buffering and double-buffering pipelines.
The implementation enables transpose-aware LDS loading for operands A and B, provided that both matrices use compatible memory layouts.
Currently, the logic performs iterations over the K dimension, while iteration over M and N dimensions is still under development and will be refined in the next update.
Future work will focus on performance evaluation and optimization of bank conflict patterns during LDS access.

## Test Plan

Basic functionality was verified using existing MFMA pipeline tests for both single and double buffering.
Next, I will extend the tests to cover various matrix layout configurations and measure execution performance.
A detailed performance table and LDS bank conflict statistics will be added in comment later to quantify the improvements.

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

stefankoncarevic committed:
[LDSTransposeLoad] Add Rock op and tests for LDS panel load
- Implemented rock.lds_transpose_load TD definition
  supporting f16 and bf16.
- Added verifier to ensure source memref is in workgroup (LDS)
  memory and indices match rank.
- Implemented lowering pattern to amdgpu.transpose_load.
- Created MLIR tests covering FP16 and BF16 loads with
  FileCheck patterns.

stefankoncarevic committed:
Add helper functions to query MFMA intrinsic K and non-K
dimensions

stefankoncarevic committed:
Implement LDS transpose load logic for MFMA pipelines
This commit introduces the full implementation of LDS transpose
load handling used in threadwise read and single-buffering
and double buffering pipelines. It adds logic for computing
per-lane base offsets, generating LdsTransposeLoadOp instructions,
and managing vectorized fragment loading for MFMA operations.
The implementation supports multiple layout kinds
(e.g., L16x16, L32x16, L32x8) and dynamically expands offsets
for multi-K fused cases.This enables more flexible data movement
between LDS and registers for MFMA input tiles.

stefankoncarevic committed:
Introduce new TOML test suite for LDS Transpose Load covering
f16 and bf16 data types, with multiple K-dimension configurations
and schedule versions.
Add CFG file to restrict execution to gfx950 architecture only,
ensuring tests run exclusively on supported hardware.
All test cases have passed validation under gfx950.

stefankoncarevic committed:
Updated tests.

stefankoncarevic committed:
Add conditional for transpose setup for A and B in single
buffering pipeline.

stefankoncarevic committed:
Refactor transpose decision handling and enabled
for double buffering. Added global decision context to propagate
hwtranspose::Decision from BlockwiseGemm to ThreadwiseReadIntoOp.
Updated ThreadwiseReadIntoOp to attach hwtranspose attributes
when a valid decision is available. Fixed double buffering
handling to ensure correct LDS access.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Fix getDecisionLdsTransposeContext for correct transpose
decision handling.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Fix attachAttributes function.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Fixed merge build issue.

stefankoncarevic committed:
Updated offset computation to load elements grouped in 16x4 blocks.
This allows correct MFMA data ordering without changing how data
is stored in LDS. Also extended iteration logic for M/N dimensions
when 64 threads are active.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
LDS transpose load distribution for multi-wave configurations

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Code style standardization:
   - Replaced all b.create<Op>() patterns with Op::create(b, loc, ...)
     for consistency with project conventions

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Replace stored panel count attributes with on-demand calculation using
block dimensions and MFMA layout parameters, reducing redundancy.

stefankoncarevic committed:
Extract M/N offset helpers in LDS transpose load

Extract three helper functions for M/N offset computation in LDS
transpose load lowering, improving code structure and readability:

- computeWaveOffset(): Wave spatial distribution offset
- computeTileIterationOffset(): Tile iteration offset (dynamic/static)
- computeFinalMNOffset(): Combines all M/N offset components

Refactor main load loop to use these helpers, replacing ~30 lines
of inline logic with clean function calls. No functional changes.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Adjust Rock lds-transpose tests to align FileCheck expectations with the new attribute names/order emitted by rock.threadwise_read_into. Ensures future pipelines consistently validate the renamed attrs without spurious diffs.

stefankoncarevic committed:
Add per-operand LDS transpose decision logic and remove unused attrs

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic commented:
With the latest changes, LDS transpose-load can be applied even when only one of the input matrices is in the expected layout. This allows us to support almost all problem configurations.

There are still some accuracy issues for a few perf configs on matrix B, and I'm currently working on identifying and resolving the root cause.

Regarding the new attributes added to blockwise_load_tile and blockwise_to_threadwise ops: additional updates will likely be required to determine the correct way to generate the IR on these ops, so that the logic is properly handled inside the threadwise_read_into stage where the all attributes is attached and required to calculate correct offset.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

pabloantoniom commented on mlir/test/Dialect/Rock/lds_transpose_attributes.mlir:
```diff
@@ -0,0 +1,68 @@
+// RUN: rocmlir-opt -split-input-file -rock-gridwise-gemm-to-blockwise \
```
Can we have this in a single line? For consistency with other tests

pabloantoniom commented on mlir/test/Dialect/Rock/lds_transpose_attributes.mlir:
```diff
@@ -0,0 +1,68 @@
+
+module attributes {mhal.arch = "amdgcn-amd-amdhsa:gfx950"} {
+  // CHECK-LABEL: func.func @test_lds_transpose_attributes_double
+  func.func @test_lds_transpose_attributes_double(
```
nit: `test_lds_transpose_attributes_double_buffering`?

pabloantoniom commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1957,6 +1957,47 @@ LogicalResult InBoundsStoreOp::verify() {
+
+LogicalResult LDSTransposeLoadOp::verify() {
+  // Source must be memref in workgroup (LDS) address space
+  MemRefType srcType = getSource().getType();
```
The code to check if the source is LDS is a bit complicated and I can see we also check this for other ops in the same file...can we create a function like `hasWorkgroupMemorySpace` in AMDGPUDialect.cpp and reuse it in here? And then do:

```
if (!hasWorkgroupMemorySpace(srcType.getMemorySpace()))
    return emitOpError("source memory address space must be Workgroup");
```

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1037 @@
+    return nullptr;
+
+  NamedAttrList attrs;
+  attrs.append("rock.lds_transpose_enabled", rewriter.getUnitAttr());
```
I understand we need `rock.lds_transpose_enabled` attribute, but, `rock.mperblock`? `rock.nperblock`? Those are in `#rock.mfma_gemm_params`, why duplicate them here?

stefankoncarevic replied:
I agree that the current approach, where attributes are introduced directly during IR generation in several places, is not ideal and leads to duplication. My plan is to move the decision logic into GridwiseGemmToBlockwise, and then propagate the required information through the pipeline so it reaches ThreadwiseReadIntoOp in a consistent way.

To do that, I’ll check which parameters are already available in #rock.mfma_gemm_params. If some parameters needed for HW transpose are missing, I’ll extend mfma_gemm_params so that all attributes remain centralized and consistent. That should address both the duplication concerns and the naming issues.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1037 @@
+    return nullptr;
+
+  NamedAttrList attrs;
+  attrs.append("rock.lds_transpose_enabled", rewriter.getUnitAttr());
```
I agree we need new new attributes for transpose loads but I don't think this is a good way of doing it. I think they should be added to `mfma_gemm_params` in `RockAttrDefs` for consistency.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -411,12 +411,18 @@ struct BlockwiseGemmAccelRewritePattern
     Value scaleA = adaptor.getBufferScaleA();
     Value scaleB = adaptor.getBufferScaleB();
+
+    bool allowHWTranspose = op->hasAttr("rock.gemm");
```
An attribute called `rock.gemm` to indicate that we want to allow transpose load? I think we can come up with a better name?

stefankoncarevic replied:
Instead of using attributes like rock.gemm as flags, I’ll introduce a clearer and more explicit attribute name that reflects the purpose.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -3142,7 +3148,8 @@ struct GridwiseGemmAccelRewritePattern
             /*bufferScaleA=*/arrayScaleA, /*bufferScaleB=*/arrayScaleB,
-            featuresAttr, op.getBlockSizeAttr(), op.getParamsAttr());
+            featuresAttr, op.getBlockSizeAttr(), op.getParamsAttr(),
+            /*enableHWTranspose=*/true);
```
Do we want to always enable transpose load? Shouldn't this be a variable coming from somewhere?

umangyadav replied:
Agree with pablo's comment. Why are enabling HW Transpose always ?

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseLoadTileToThreadwise.cpp:
```diff
@@ -193,6 +203,29 @@ class LoweringBlockwiseLoadTileOp final
+    // Decide per operand whether LDS transpose can be applied. We do this
+    // locally so that each lowered ThreadwiseReadIntoOp carries only its own
+    // attributes.
+    DictionaryAttr transposeAttr = nullptr;
```
This is similar to the `buildTransposeAttr` lambda in BlockwiseGemmToThreadwise.cpp. Can we reuse code?

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,108 @@
+};
+
+// Structure to hold the outcome of the hardware transpose analysis.
+struct Decision {
```
Could this be merged with LoweringInfo?

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,108 @@
+
+// Lowering-time description.
+struct LoweringInfo {
+  bool usable{false};
```
What exactly is "usable" here? Can you add documentation?

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1037 @@
+        "destination capacity");
+  }
+
+  b.replaceOp(op, ValueRange{});
```
Do you want b.eraseOp?

stefankoncarevic replied:
Thanks for pointing that out, so I’ve switched the final builder call to b.eraseOp(op).

pabloantoniom reviewed (commented):
Leaving some high-level comments for now. I want to take a deeper look at `emitThreadwiseHWTranspose`, I'll post more comments once I get some time to do it.

stefankoncarevic committed:
Fix and refactor wave grid layout computation for LDS transpose load

stefankoncarevic committed:
Changed replaceOp with eraseOp.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseLoadTileToThreadwise.cpp:
```diff
@@ -157,6 +163,10 @@ class LoweringBlockwiseLoadTileOp final
 
+    bool allowHWTranspose = op->hasAttr("rock.gemm");
+    if (allowHWTranspose)
+      op->removeAttr("rock.gemm");
```
I agree with Pablo's earlier comment that this needs a better name, but also why does it need explicit removal here?

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,108 @@
+  OperandKind operand{OperandKind::A};
+  Type elemType;
+  bool destIsVector{false};
+  VectorType destVecType;
```
It looks like based on the usage in LdsTransposeLoad.cpp that these fields are dead (i.e., never get used). Can they be removed?

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,108 @@
+  bool usable{false};
+  LayoutKind layout{LayoutKind::None};
+  OperandKind operand{OperandKind::A};
+  Type elemType;
```
Nit: Could we add a default initializer for elemType here so that it is consistent with the other struct members. Something like `Type elemType{}`?

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/ThreadwiseGemmLowering.cpp:
```diff
@@ -796,6 +796,22 @@ LogicalResult ThreadwiseReadIntoRewritePattern::matchAndRewrite(
   auto globalToLDSTransform = maybeGlobalToLDSTransform.value();
 
+  // Check if the operation has the attribute for LDS Transpose Load
+  if (op->hasAttr("rock.lds_transpose_enabled")) {
```
Do we want to add the attribute to `ThreadwiseReadIntoO` so that we don't have to call `op->hasAttr`?

umangyadav replied:
Yeah that would be better. 
Or else would it be possible to check the input and output shapes along with rock.transform sequence on the threadwise_read_into_op to determine if the threadwise_read_into_op is doing a transpose or not ?

justinrosner commented on mlir/test/e2e/PrLdsTransposeLoad.toml:
```diff
@@ -0,0 +1,57 @@
+directory = "PrLdsTransposeLoad"
+prefix = "rocmlir-gen"
+suffix = "--operation gemm --arch %arch %pv %random_data %rocmlir_gen_flags | rocmlir-driver -c | mlir-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_float16_utils%shlibext --entry-point-result=void | FileCheck %s -- ...
```
We can sometimes run into intermittent accuracy issues on the random E2E tests when using unconstrained input ranges with bf16 data types. Maybe look to use `%constrained_float_range_random_data` that was introduced in https://github.com/ROCm/rocMLIR/pull/2116

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1085 @@
+    return dec;
+  }
+  if (!(elemTypeA.isF16() || elemTypeA.isBF16()) ||
+      !(elemTypeB.isF16() || elemTypeB.isBF16())) {
```
We already check that `elemTypeA` is equal to `elemTypeB` in the previous if statement. So there is no need to have the check for both here.

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1085 @@
+}
+
+StringRef layoutName(LayoutKind kind) {
+  for (const auto &config : kLayoutConfigs)
```
Nit: Can you add braces around the for loop here. It makes it consistent with how it is done in the rest of the file, and improves readability slightly.

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1085 @@
+  auto destType = cast<MemRefType>(dest.getType());
+  Type elemType = info.elemType;
+  Value sourceView = op.getSource();
+  auto [rawSrc, _, __] = untransform(b, sourceView);
```
Nit: Add comments explaining what the unused values are. E.g., `[rawSrc, /*transforms*/ _, ...`

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1085 @@
+  }
+  case LayoutKind::L32x16: {
+    // mbase = mOffsetBase + (blockId % 2) * 16
+    Value c16_2 = arith::ConstantIndexOp::create(b, loc, 16);
```
Can we just reuse `c16` here?

stefankoncarevic replied:
done.

justinrosner commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1085 @@
+        kOffsetBase);
+
+    // m_offset_base = mOffsetBase + (blockId % 2) * 16
+    Value c16_2 = arith::ConstantIndexOp::create(b, loc, 16);
```
Same here?

stefankoncarevic replied:
done.

stefankoncarevic committed:
Route the LDS checks through mlir::rock::hasWorkgroupMemorySpace() so both AddressSpaceAttr and integer encodings are handled consistently.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Address review feedbacky

stefankoncarevic committed:
Address review feedback

stefankoncarevic committed:
Address review feedback

umangyadav commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,29 @@ defvar SameShapeVectorOfI1 = [{
+    LDS transpose panel load (vector<4 x F16/BF16>) for MFMA-friendly ordering.
+    Loads a single panel (4 elements) from LDS memory using AMD ds.read.tr16.* instructions.
+    The returned fragment ($fragment) is always a vector<4 x elementType>. 
+    Helps to reduce bank conflict. Only supported for gfx950.
```
It is better to not mention which instruction it lowers to or the fact that it will readh `vector<4xf16>`.  We could use this instruction in future for some other dtype or some other vector length as well.

stefankoncarevic replied:
done.

umangyadav commented on mlir/include/mlir/Dialect/Rock/IR/Rock.h:
```diff
@@ -63,6 +63,10 @@ ArrayAttr noTransformsArray(Builder &b, size_t n);
 
+// Returns true if the provided memory space attribute encodes GPU workgroup
+// memory.
+bool hasWorkgroupMemorySpace(Attribute memorySpace);
```
```suggestion
bool isWorkgroupMemorySpace(Attribute memorySpace);
```

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1957,6 +1958,33 @@ LogicalResult InBoundsStoreOp::verify() {
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  // This op only reads from LDS (workgroup) memory
+  auto *read = MemoryEffects::Read::get();
+  effects.emplace_back(read, &getSourceMutable());
```
Doesn't also write back transposed matrix into VGPRs ?  It should have write effect on the destination.

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1957,6 +1958,33 @@ LogicalResult InBoundsStoreOp::verify() {
+    return emitOpError("source memory address space must be workgroup (LDS)");
+
+  // Indices size must match rank
+  if (getIndices().size() != srcType.getRank())
```
Can we check that this op only appears on the hardware that supports it ?

stefankoncarevic replied:
done.

umangyadav commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,29 @@ defvar SameShapeVectorOfI1 = [{
   ::mlir::VectorType::get(::llvm::cast<::mlir::ShapedType>($_self).getShape(), ::mlir::IntegerType::get($_ctxt, 1))
 }];
 
+// lds_transpose_load
```
Can you add some tests in `rock/ops.mlir` and `rock/ops_error.mlir` to make sure this operation is being constructed correctly and verified correctly ?

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1957,6 +1958,33 @@ LogicalResult InBoundsStoreOp::verify() {
+void LDSTransposeLoadOp::getEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  // This op only reads from LDS (workgroup) memory
+  auto *read = MemoryEffects::Read::get();
```
add tests in `rock/effects.mlir` for testing the effects.

stefankoncarevic replied:
done.

umangyadav commented on mlir/include/mlir/Dialect/Rock/IR/AccelEmitter.h:
```diff
@@ -202,6 +202,12 @@ struct MfmaEmitter : public AccelEmitter {
 
   int64_t getRowGroupSize() const;
 
+  // Return the MFMA K dimension
```
nit: 

Can we move this into base class AccelEmitter and override it ? So that we can extend this to wmma in future.

stefankoncarevic replied:
done.

umangyadav commented on mlir/include/mlir/Dialect/Rock/IR/Rock.h:
```diff
@@ -63,6 +63,10 @@ ArrayAttr noTransformsArray(Builder &b, size_t n);
 
+// Returns true if the provided memory space attribute encodes GPU workgroup
+// memory.
+bool hasWorkgroupMemorySpace(Attribute memorySpace);
```
Can we move this inside `rock/utility/loweringUtils.h` instead ?

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1080 @@
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "rock-hw-transpose-support"
```
Keep this debug name similar to pass name `rock-lds-transposed-load`

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,107 @@
+
+// The main decision-making function. It analyzes the GEMM parameters and
+// returns a Decision struct indicating if the optimization is applicable and
+// with which paneling configuration.
```
Is the term paneling same as the `vector` ? I've not see panel term being used before.

stefankoncarevic replied:
done, it was older comment.

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.h:
```diff
@@ -0,0 +1,107 @@
+LayoutKind selectLayout(int64_t nonKDim, int64_t instrK);
+
+// Attach attributes to the ThreadwiseReadIntoOp based on the decision.
+DictionaryAttr buildTransposeAttr(const Decision &dec, bool isOperandA,
```
nit: 
Keep `PatternRewriter` as the first argument for the all function. Followed by `location` if it exists.

stefankoncarevic replied:
done.

umangyadav commented on mlir/test/Dialect/Rock/load_transpose_lds.mlir:
```diff
@@ -0,0 +1,17 @@
+// RUN: rocmlir-opt --rock-sugar-to-loops %s | FileCheck %s 
```
For some reason i've seen prefix `lowering_` for the `rock-sugar-to-loops` tests. Can we continue using that ?

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -144,19 +147,22 @@ static void loadAndStoreGemmInputTile(
                  gridCoords.n_block, tid},
       featuresAttr, rewriter.getI32IntegerAttr(blockSize), gemmTuningParams);
+  if (enableHWTranspose)
+    blockwiseLoad->setAttr("rock.gemm", rewriter.getUnitAttr());
```
What this unitAttr is about ? and why is it called `rock.gemm` ?

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -109,11 +110,13 @@ static void blockwiseGemmAccel(
+      blockSize, params);
+  if (enableHWTranspose)
+    gemmOp->setAttr("rock.gemm", rewriter.getUnitAttr());
+  return gemmOp;
```
Why is it returning blockWiseGemmOp ? I do not find return value from this function being used anywhere.

stefankoncarevic replied:
done.

umangyadav commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseLoadTileToThreadwise.cpp:
```diff
@@ -193,6 +203,29 @@ class LoweringBlockwiseLoadTileOp final
+    if (allowHWTranspose) {
+      if (auto *mfma =
+              dyn_cast<rock::accel::MfmaEmitter>(accelEmitterPtr.get())) {
+        if (!matrixParams.getLDSLayoutDxK()) {
```
I think it should check if it reading from LDS or not. and then assert that if it is reading from LDS then it should have a valid LDS Layout.

umangyadav reviewed (commented):
When we lower blockwise_load_tile, it can emit threadwise_copy ops that does the transposes if necessary. Instead can we get rid of those with transposed_load from LDS on gfx950 ?

stefankoncarevic committed:
Refactor: Extract stride calculation and fix multi-wave logic

Previously, a single "mnStride" parameter caused 50% accuracy failures in
multi-wave scenarios because it conflated two distinct concepts:
- Spatial wave separation (should always be nonKDim)
- Outer loop tile iteration step (depends on wave interleaving)

This commit splits the logic into two parameters:
- waveOffsetStride: ALWAYS nonKDim (wave separation)
- tileOffsetStride: wavesInDim × nonKDim or nonKDim (tile iteration)

Changes:
- Add StrideConfig struct with waveOffsetStride and tileOffsetStride
- Create computeStrideConfiguration() helper with proper documentation
- Update emitThreadwiseHWTranspose() to use the new two-stride approach

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Fixed review feedback

stefankoncarevic committed:
Address review feedback

stefankoncarevic committed:
Address review feedback

stefankoncarevic committed:
Add LDS transpose configuration attributes and refactor decision logic

Introduce Rock_LdsTransposeConfigAttr to store MFMA geometry and tiling
parameters for hardware-assisted LDS transpose optimization. Extend
MfmaGemmParamsAttr with optional mfmaNonKDim/mfmaKDim fields and
BlockwiseMatrixParamsAttr with ldsTransposeEnabled flag.

Refactor decision-making logic to GridwiseGemmToBlockwise pass, which
now decides enablement for both operands globally and propagates the
decision downstream via attributes. BlockwiseGemmToThreadwise and
BlockwiseLoadTileToThreadwise passes consume these attributes instead
of re-running eligibility checks.

Move internal helpers to anonymous namespace in LdsTransposeLoad.cpp
for better encapsulation. Replace intermediate layout enums with direct
use of mfmaNonKDim/mfmaKDim values to reduce indirection.

Add parsing/printing tests in ops.mlir and integration tests in
lds_transpose_attributes.mlir covering single/double buffering paths.

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic commented:
Refactors the LDS transpose optimization to use structured attributes for decision propagation, eliminating redundant eligibility checks and improving maintainability.
**1. GridwiseGemmToBlockwise.cpp (Decision Point)**
- Calls `decideLdsTransposeForOperands()` once for both operands A and B
- Applies kpack constraint: only enable for single operand if `kpack == 1`
- Sets `ldsTransposeEnabled` flag in `BlockwiseMatrixParamsAttr` for each operand
- If enabled, populates `mfmaNonKDim`/`mfmaKDim` in `MfmaGemmParamsAttr`

**IR Example (after GridwiseGemmToBlockwise):**
rock.blockwise_gemm_accel %C += ... features = mfma|dot {
  matrixParamsA = #rock.blockwise_matrix_params<
    ..., ldsTransposeEnabled = true>,
  params = #rock.mfma_gemm_params<
    ..., mfmaNonKDim = 32, mfmaKDim = 16>
}---

#### **2. BlockwiseGemmToThreadwise.cpp (Single Buffering Path)**
- Checks `matrixParams.getLdsTransposeEnabled()` for each operand
- If true, extracts `mfmaNonKDim`/`mfmaKDim` from `MfmaGemmParamsAttr`
- Calls `buildTransposeAttrFromParams()` to create `LdsTransposeConfigAttr`
- Attaches attribute to `ThreadwiseReadIntoOp`

**IR Example (after BlockwiseGemmToThreadwise):**
rock.threadwise_read_into {
  forceUnroll,
  ldsTransposeConfig = #rock.lds_transpose_config<
    mfmaNonKDim = 32, mfmaKDim = 16,
    mPerBlock = 64, nPerBlock = 64, kPerBlock = 16,
    mPerWave = 32, nPerWave = 32,
    doubleBuffering = false, isOperandA = true>,
  useIndexDiffs
} [](%ldsBuffer) [%tid, %loopVar] -> %regs---

#### **3. BlockwiseLoadTileToThreadwise.cpp (Double Buffering Path)**
- Same logic as single buffering: reads `ldsTransposeEnabled` → builds `LdsTransposeConfigAttr`
- Attaches to `ThreadwiseReadIntoOp` inside double-buffered load loops

---

#### **4. ThreadwiseGemmLowering.cpp (Final Lowering)**
- Checks if `ThreadwiseReadIntoOp` has `ldsTransposeConfig` attribute
- If present, calls `emitThreadwiseHWTranspose()` which:
  - Extracts MFMA geometry (`mfmaNonKDim`, `mfmaKDim`) directly from attribute
  - Computes wave grid layout, panel offsets, and stride configuration
  - Emits `rock.lds_transpose_load` operations for hardware execution
  
 #### **Rock_LdsTransposeConfigAttr (RockAttrDefs.td)**
def Rock_LdsTransposeConfigAttr : AttrDef<RockDialect, "LdsTransposeConfig"> {
  let parameters = (ins
    "int64_t":$mfmaNonKDim,    // MFMA non-K dimension (16 or 32)
    "int64_t":$mfmaKDim,       // MFMA K dimension (8, 16, or 32)
    "int64_t":$mPerBlock,      // Tiling parameters
    "int64_t":$nPerBlock,
    "int64_t":$kPerBlock,
    "int64_t":$mPerWave,
    "int64_t":$nPerWave,
    "bool":$doubleBuffering,   // Single vs double buffering
    "bool":$isOperandA         // Operand A or B
  );
}
#### **Extended `MfmaGemmParamsAttr`:**
OptionalParameter<"int64_t">:$mfmaNonKDim,  // NEW
OptionalParameter<"int64_t">:$mfmaKDim      // NEW
#### **Extended `BlockwiseMatrixParamsAttr`:**
DefaultValuedParameter<"bool", "false">:$ldsTransposeEnabled  // NEW

### 🧪 **Testing:**

- **ops.mlir**: Parsing/printing tests for all 4 MFMA geometries (16x16, 16x32, 32x8, 32x16)
- **lds_transpose_attributes.mlir**: Integration tests for single/double buffering paths

stefankoncarevic committed:
Address review feedback

stefankoncarevic committed:
Please enter the commit message for your changes. Lines starting
Fix clang format

stefankoncarevic committed:
Add and fix tests.

stefankoncarevic committed:
Updated perf configs

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -3028,10 +3057,29 @@ struct GridwiseGemmAccelRewritePattern
+
+    // Create enhanced params with MFMA geometry (only if at least one operand
+    // can use it)
+    RockAccelTuningParamAttrInterface enhancedParams =
```
Instead of calling a function that returns a new createEnhancedTuningParams with the new info, could we just modify tuningParams instead?

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -436,9 +439,44 @@ struct BlockwiseGemmAccelRewritePattern
+      bool loadFromLDS = isOperandA ? loadAFromLDS : loadBFromLDS;
+
+      // Check if LDS transpose is enabled for this operand
+      if (!loadFromLDS || !matrixParams.getLdsTransposeEnabled())
```
Is it enough to check for `!matrixParams.getLdsTransposeEnabled()`. In other words, if `matrixParams.getLdsTransposeEnabled()` shouldn't be true if `!loadFromLDS`

stefankoncarevic replied:
done, I removed the redundant !loadFromLDS check.

pabloantoniom commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -509,7 +502,38 @@ def Rock_BlockwiseMatrixParamsAttr : Rock_Attr<"BlockwiseMatrixParams", []> {
+    rock.lds_transpose_load operations with the correct MFMA geometry
+    and tiling parameters.
+    
+    - mfmaNonKDim: MFMA instruction non-K dimension (16 or 32)
```
I'm not sure about my own suggestion but should this be mfmaDDim? Since we call D for M or N across the codebase?

stefankoncarevic replied:
done, changed to mfmaDDim.

pabloantoniom commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+    Arguments<(ins Arg<MemRefOf<[F16, BF16]>, "LDS source buffer">:$source,
+        Variadic<Index>:$indices)>,
+    Results<(outs VectorOfLengthAndType<[4], [F16, BF16]>:$fragment)> {
+  let summary = "Hardware-assisted LDS transpose panel load (v4 fragment)";
```
What does v4 fragment mean?

stefankoncarevic replied:
The instruction always returns a vector of **4 elements** (`vector<4xf16>` or 
`vector<4xbf16>`)

pabloantoniom commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1233,9 +1256,16 @@ def Rock_ThreadwiseReadIntoOp
-        build($_builder, $_state, TypeRange{}, source, dest, ValueRange{}, extraViews, extraIndices, forceUnroll, useIndexDiffs);
-      }]>
-  ];
+        build($_builder, $_state, TypeRange{}, source, dest, ValueRange{}, extraViews, extraIndices, forceUnroll, useIndexDiffs, /*ldsTransposeConfig=*/nullptr);
```
Because of this builder, it shouldn't be necessary to change the `ThreadwiseReadIntoOp::create` calls to explicitly pass nullptr right? See `BlockwiseGemmToThreadwise` for examples.

stefankoncarevic replied:
thanks, done.

pabloantoniom commented on mlir/test/Dialect/Rock/load_transpose_lds.mlir:
```diff
@@ -0,0 +1,17 @@
+// RUN: rocmlir-opt --rock-sugar-to-loops %s | FileCheck %s 
+
+// CHECK-LABEL: func @test_load_transpose_fp16
+module attributes {mhal.arch = "amdgcn-amd-amdhsa:gfx950"} {
```
Can we add another test with another arch (i.e., gfx942) to ensure that lowering fails with an error message saying only gfx950 is supported?

pabloantoniom commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1990,6 +1992,45 @@ LogicalResult InBoundsStoreOp::verify() {
+      return emitOpError("indices must be of index type");
+  }
+
+  return success();
```
I think we also need to check that the datatype match what we expect (i.e., typeA == typeB && typeA is f16 or bf16).

Also adding a test in ops_error.mlir would be nice.

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/test/Dialect/Rock/ops_error.mlir:
```diff
@@ -1162,3 +1162,53 @@ func.func @global_load_to_lds_f4_dest_odd_coord(
+  %fragment = rock.lds_transpose_load %buffer[%c0, %c0, %c0]
+    : memref<128x64xf16, #gpu.address_space<workgroup>> -> vector<4xf16>
+  return
+}
```
I would also add some test/s to make sure `isValidMfmaGeometry` is working properly.

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+  // Check basic constraints for LDS transpose applicability
+  // - directToLDS must be enabled
+  // - Layout must be KxD (not DxK)
+  bool canUseForA = directToLDS && !ldsLayoutConfigA.ldsLayoutDxK;
```
Here we check basic constraints that, if are not met, we cannot use transpose loads. Therefore, should we move this into makeDecision, where all those checks are?

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+                              << (decA.usable ? "operand A" : "operand B")
+                              << " only (kpack=1)\n");
+    } else {
+      // kpack > 1 with asymmetric support - disable both (current limitation)
```
Maybe we can open a ticket to track this feature?

stefankoncarevic replied:
done, ticket number is 2115 on internal repo.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+
+  // IMPORTANT: Only set MFMA geometry if at least one operand will use it
+  // This avoids polluting tuning params with unnecessary data
+  if (result.enableA || result.enableB) {
```
nit: I think this is equivalent to an else of the previous conditions?

stefankoncarevic replied:
changed, thanks.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+                          << "\n");
+
+  // Each panel vector contains 4 elements (ds_read_tr16_b64 → vector<4xf16>)
+  constexpr int64_t elementsPerVector = 4;
```
nit: Could we get this from somewhere instead of hardcoding? i.e., dest.getElemType().getShape() or something like that? And then asserting is 4

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+    // 16x32 layout
+    kOffsetBase = arith::MulIOp::create(b, loc, blockId, c8);
+  } else {
+    return nullptr; // Should not happen for double-rate
```
nit: llvm_unreachable?

stefankoncarevic replied:
done.

pabloantoniom commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
```diff
@@ -0,0 +1,1228 @@
+  return finalOffset;
+}
+
+LogicalResult emitThreadwiseHWTranspose(PatternRewriter &b,
```
Overall I think the doc in this PR is excellent, but the big guy here does not have any explanation. Can you add a comment to explain the high-level steps that emitThreadwiseHWTranspose does, like a step-by-step walktrough?

stefankoncarevic replied:
added

pabloantoniom reviewed (commented):
I think I left a lot of comments, but overall this looks good to me. I feel like the code overall is quite clean, easy to follow and well documented so that makes reviewing much better despite being a big PR - good job!

stefankoncarevic committed:
Fix clang format

stefankoncarevic committed:
Addressed review comments

stefankoncarevic committed:
Addressed all review feedback

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

stefankoncarevic committed:
Add gfx950 architecture guard for LDS transpose E2E tests

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/AccelEmitter.h:
```diff
@@ -150,6 +150,18 @@ struct AccelEmitter {
   AccelEmitterParams getParams() const { return accelEmitterParams; }
 
+  // Return the accelerator K dimension
+  virtual int64_t getMfmaK() const {
```
you can just do `virtual int64_t getMfmaK() const = 0`

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/AccelEmitter.h:
```diff
@@ -150,6 +150,18 @@ struct AccelEmitter {
+  }
+
+  // Return the accelerator instruction's non-K dimension
+  virtual int64_t getMfmaNonKDim() const {
```
same here

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/MfmaInsnGroup.h:
```diff
@@ -36,13 +36,13 @@ enum class MfmaTypeId : uint32_t {
 
 struct MfmaInsnInfo {
   MfmaTypeId type;
-  int64_t mfmaNonKDim;
```
why do we need to change this name?

stefankoncarevic replied:
Renamed getMfmaNonKDim() → getMfmaDDim() for consistency naming across the codebase, more intuitive naming "D dimension" is clearer than "non-K dimension", because we have a lot spot that we say d dimension, so just to be same here as well.

dhernandez0 replied:
fair enough

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -321,19 +321,12 @@ def Rock_MfmaGemmParamsAttr
+      "int64_t":$nPerBlock, "int64_t":$kpack, "int64_t":$mPerWave,
+      "int64_t":$nPerWave, "int64_t":$mnPerXdl, "int64_t":$splitKFactor,
+      "int64_t":$scheduleVersion, "int64_t":$outputSwizzle, "bool":$forceUnroll,
+      OptionalParameter<"::std::optional<int64_t>">:$mfmaDDim,
```
why do we need to add tuning params?

dhernandez0 replied:
mfmaDDim is just mnPerXdl, no?
and `mfmaKDim` is not something we tune, so I don't think it should be here.

stefankoncarevic replied:
The formula is: mnPerXdl = mfmaDDim × blocksMfma
I think for currently supported MFMA instructions they are the same, instructions have blocksMfma=1, we could derive mfmaDDim from mnPerXdl. But using mnPerXdl directly is cleaner. I agree we can use mnPerXdl instead of adding mfmaDDim to the params. 
I will add mfmaKDim to BlockwiseMatrixParamsAttr instead of MfmaGemmParams. 
I need this to be attach to thredwisereadintoOp to have all those attributes to calculate correct offset.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -509,11 +502,41 @@ def Rock_BlockwiseMatrixParamsAttr : Rock_Attr<"BlockwiseMatrixParams", []> {
+    rock.lds_transpose_load operations with the correct MFMA geometry
+    and tiling parameters.
+    
+    - mfmaDDim: MFMA instruction D dimension (M or N, 16 or 32)
```
this will be needed for gfx1250. Can we make it not reference mfma?

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -509,11 +502,41 @@ def Rock_BlockwiseMatrixParamsAttr : Rock_Attr<"BlockwiseMatrixParams", []> {
+    
+    - mfmaDDim: MFMA instruction D dimension (M or N, 16 or 32)
+    - mfmaKDim: MFMA instruction K dimension (8, 16, or 32)
+    - mPerBlock: M dimension size per block
```
all this info is already in `BlockwiseMatrixParamsAttr`, could we use it instead?

stefankoncarevic replied:
LdsTransposeConfigAttr contains additional parameters that 
BlockwiseMatrixParamsAttr doesn't have:
- Block dimensions: mPerBlock, nPerBlock, kPerBlock
- Wave dimensions: mPerWave, nPerWave  
- doubleBuffering mode
- isOperandA flag

These are all needed by threadwise_read_into to generate correct LDS transpose 
load indexing. The LdsTransposeConfigAttr is designed to be a self-contained 
configuration for this specific operation, avoiding the need to pass multiple 
separate attributes and unrelated attributes for lds transpose load.

dhernandez0 replied:
but all that info is in the tuning params, don't we have access to that as well?

stefankoncarevic replied:
Yes, the tuning params contain this info, and higher-level ops like BlockwiseGemmOp and BlockwiseLoadTileToThreadwiseOp have access to both BlockwiseMatrixParamsAttr and tuning parameters. However, when we lower to ThreadwiseReadIntoOp, those attribute groups are not attached to the op.
Rather than propagating entire attribute groups (which would include many unrelated fields like rotateDWithK, swapThreadIterSubDims, splitKAcrossThreadsFirst, etc.), I created a focused LDSTransposeConfigAttr containing only the specific parameters needed for LDS transpose load indexing. This keeps the op's interface clean and avoids cluttering the IR with unused attributes.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+def Rock_LDSTransposeLoadOp
+    : Rock_Op<"lds_transpose_load", [DeclareOpInterfaceMethods<
+                                        MemoryEffectsOpInterface>]>,
+      Arguments<(ins Arg<MemRefOf<[F16, BF16]>, "LDS source buffer">:$source,
```
did you implement bf16 as well here? I thought this was f16 only

stefankoncarevic replied:
Yes, it's also have support for bf16.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+      Results<(outs VectorOfLengthAndType<[4], [F16, BF16]>:$fragment)> {
+  let summary = "Hardware-assisted LDS transpose panel load (v4 fragment)";
+  let description = [{
+    LDS transpose panel load for MFMA-friendly ordering.
```
what is `panel`?

stefankoncarevic replied:
"Panel" refers to a tile of data whose dimensions match the selected matrix-multiply accelerator instruction geometry.
Specifically:
Panel dimensions = dDim × instrK (e.g., 16×16, 32×8, 32×16, or 16×32)
The block is divided into multiple panels (mPerBlock / dDim panels in M dimension, kPerBlock / instrK panels in K dimension). Each panel represents the granularity at which the hardware transpose load operates - one or more ds_read_tr16_b64 instructions are used to load a complete panel, depending on the K dimension.
I'll update the description to clarify this.

stefankoncarevic replied:
done

dhernandez0 replied:
thanks for adding this info. nit: I'd prefer wmma/mfma tile size instead of panel, it seems more intuitive

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+      Arguments<(ins Arg<MemRefOf<[F16, BF16]>, "LDS source buffer">:$source,
+          Variadic<Index>:$indices)>,
+      Results<(outs VectorOfLengthAndType<[4], [F16, BF16]>:$fragment)> {
+  let summary = "Hardware-assisted LDS transpose panel load (v4 fragment)";
```
what is `v4 fragment`?

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+  let description = [{
+    LDS transpose panel load for MFMA-friendly ordering.
+    Loads a panel from LDS memory with transpose support.
+    The returned fragment ($fragment) contains elements in transposed layout suitable for MFMA operations.
```
what is `fragment`? I would suggest defining any term you are using for the first time.

dhernandez0 replied:
also, I'd prefer to avoid mentioning mfma (because of gfx1250)

stefankoncarevic replied:
v4 fragment refers to the vector-of-4 result format (vector<4xf16> or vector<4xbf16>) that the ds_read_tr16_b64 hardware instruction returns per thread. Each thread in the wave receives 4 consecutive transposed elements.

stefankoncarevic replied:
done.

dhernandez0 replied:
instead of saying fragment can we say vector of 4?

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -1141,6 +1141,28 @@ defvar SameShapeVectorOfI1 = [{
+    LDS transpose panel load for MFMA-friendly ordering.
+    Loads a panel from LDS memory with transpose support.
+    The returned fragment ($fragment) contains elements in transposed layout suitable for MFMA operations.
+    Helps to reduce bank conflicts. Only supported for gfx950.
```
`Only supported for gfx950` I'd avoid this here, otherwise we might forget to update it in the future

stefankoncarevic replied:
I will update description with this couple suggested changes and have better description.

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -74,6 +74,20 @@ using namespace mlir::rock;
 //===----------------------------------------------------------------------===//
 
+bool mlir::rock::isWorkgroupMemorySpace(Attribute memorySpace) {
+  if (!memorySpace)
```
we assume if no memorySpace is not workgroup, I'd rather fail (FailureOr)

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1990,6 +2034,56 @@ LogicalResult InBoundsStoreOp::verify() {
+  StringAttr archAttr =
+      rock::getArch(*this).value_or(StringAttr::get(getContext(), "gfx00"));
+  StringRef arch = archAttr.getValue();
+  if (!arch.contains("gfx950")) {
```
I think we wanted to have all hw specific checks in AmdArchDb?

dhernandez0 replied:
also, are we sure we want to have hw checks in verify? we don't do that for other cases

stefankoncarevic replied:
I've moved the hardware check to AmdArchDb (added `hasLdsTransposeLoad` 
field), so now it uses `lookupArchInfo(arch).hasLdsTransposeLoad` instead of 
hardcoded string comparison. The verify check was added per Umang's request for early error detection. I can 
remove it if we prefer to handle this during lowering instead.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -436,9 +439,43 @@ struct BlockwiseGemmAccelRewritePattern
+      return hwtranspose::buildTransposeAttrFromParams(
+          b, mfmaDDim.value(), mfmaKDim.value(), mPerBlock, nPerBlock,
+          kPerBlock, mPerWave, nPerWave,
+          /*doubleBuffering=*/false, isOperandA);
```
why is `doubleBuffering=false`?

dhernandez0 replied:
ah makes sense, because if we lower this, we are doing single buffer. Please add a comment about that

stefankoncarevic replied:
done

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -492,10 +529,10 @@ struct BlockwiseGemmAccelRewritePattern
+    auto loadBuffer =
+        [&](Value buffer, Value wrappedLDSBufferForLoad, Value loopVar,
+            Type argType, int64_t repeats, bool loadFromLDS, bool directToLDS,
+            bool isA, LdsTransposeConfigAttr transposeAttr = nullptr) -> Value {
```
if we are always passing `transposeAttr`, do not set a default value.

stefankoncarevic replied:
The default value = nullptr is necessary because scale buffers not yet support for fp4.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseLoadTileToThreadwise.cpp:
```diff
@@ -75,7 +76,8 @@ class LoweringBlockwiseLoadTileOp final
       Value tid, StringRef dName, Value ldsView, Value regs, int64_t blockSize,
-      bool forceUnroll, const BlockwiseMatrixParamsAttr &matrixParams) const {
+      bool forceUnroll, const BlockwiseMatrixParamsAttr &matrixParams,
+      LdsTransposeConfigAttr transposeAttr = nullptr) const {
```
if we are always passing `transposeAttr`, do not set a default value.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -2276,25 +2276,29 @@ struct GridwiseAttentionAccelRewritePattern
         ldsLayoutCfgMG0.ldsLayoutDxK, directToLDS,
-        /*splitKAcrossThreadsFirst=*/false, gemm0G, gemm0M, gemm0InMPerThread);
+        /*splitKAcrossThreadsFirst=*/false, gemm0G, gemm0M, gemm0InMPerThread,
+        /*ldsTransposeEnabled=*/false);
```
why is it false for attention?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -3028,10 +3032,39 @@ struct GridwiseGemmAccelRewritePattern
 
+    // ============================================================
+    // LDS TRANSPOSE DECISION MAKING
+    // ============================================================
```
can we move all of this to where it's used? (ThreadwiseReadInto lowering)

stefankoncarevic replied:
I considered moving the decision logic to ThreadwiseReadInto lowering, but 
it would require passing additional information through attributes (arch, 
ldsLayoutConfig, doubleBuffering, etc.) that isn't currently available there.
The current approach keeps the decision where all required information is 
accessible, and only propagates the result (ldsTransposeEnabled, accelKDim) 
through BlockwiseMatrixParamsAttr.
This follows the pattern of making decisions early when information is 
available, then passing minimal results downstream.

Happy to refactor if you see a cleaner approach, but I think the current 
design minimizes attribute complexity while keeping the logic centralized.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/LdsTransposeLoad.cpp:
this doesn't implement any pass, does it? I think it should go into utils, not Transforms.

stefankoncarevic replied:
done.

stefankoncarevic committed:
- Rename getMfmaNonKDim() → getMfmaDDim() for clarity
- Make getMfmaK() and getMfmaDDim() pure virtual in AccelEmitter
- Add WmmaEmitter implementations deriving values from wmmaInsn
- Update documentation to use "accelerator" instead of "MFMA"
- Define "panel" terminology in LDS transpose docs

stefankoncarevic committed:
Addressed review comments

stefankoncarevic committed:
Move LDS transpose helpers from Transforms to utility

stefankoncarevic committed:
Added new files location.

stefankoncarevic committed:
Move LDS transpose geometry to BlockwiseMatrixParamsAttr

- Remove mfmaDDim/mfmaKDim from MfmaGemmParamsAttr (not tuning params)
- Add accelKDim to BlockwiseMatrixParamsAttr
- Use mnPerXdl for accelDDim in LDS transpose config
- Add hasLdsTransposeLoad to AmdArchDb for hw capability check
- Replace hardcoded "gfx950" check with lookupArchInfo()

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/utility/LdsTransposeLoad.h:
nit: Lds -> LDS?

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -1330,12 +1386,13 @@ LogicalResult LiveInOp::verify() {
+  if (failed(memSpaceCheck))
+    return emitError("The operand of rock.live_in must have a specified "
+                     "memory space");
+  if (!*memSpaceCheck)
```
nit: memSpaceCheck.value()

stefankoncarevic replied:
done.

dhernandez0 commented on mlir/test/e2e/PrLdsTransposeLoad.cfg:
```diff
@@ -0,0 +1,2 @@
+if not (hasattr(config, 'arch') and config.arch and "gfx950" in config.arch):
```
nit: use has_lds_tranpose... flag

stefankoncarevic replied:
done.

dhernandez0 reviewed (approved):
minor nits but looks good, thanks!

stefankoncarevic committed:
Address LDS transpose review feedback- Rename Lds -> LDS for naming consistency- Use memSpaceCheck.value() instead of *memSpaceCheck- Use lds_transpose_load feature flag instead of hardcoded gfx950 check- Rename $fragment -> $result in op definition- Replace "panel" with "matrix accelerator tile" terminology- Make descriptions hardware-agnostic (remove MFMA-specific wording)

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16

pabloantoniom reviewed (approved):
Great job!

justinrosner commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -321,19 +321,11 @@ def Rock_MfmaGemmParamsAttr
-    "int64_t":$outputSwizzle,
-    "bool":$forceUnroll
-  );
+  let parameters = (ins "int64_t":$kpackPerBlock, "int64_t":$mPerBlock,
+      "int64_t":$nPerBlock, "int64_t":$kpack, "int64_t":$mPerWave,
+      "int64_t":$nPerWave, "int64_t":$mnPerXdl, "int64_t":$splitKFactor,
+      "int64_t":$scheduleVersion, "int64_t":$outputSwizzle,
+      "bool":$forceUnroll);
```
Nit: I'm slightly in favor of having one input parameter per line in TableGen files. Additionally, this tends to be the way that it's done in upstream

justinrosner reviewed (approved):
Looks good!

stefankoncarevic committed:
Merge branch 'develop' into lds_transpose_load_ds_read_tr16
